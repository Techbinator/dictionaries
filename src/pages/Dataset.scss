.dataset {
  display: flex;
  .table-wrapper,
  .filter-wrapper {
    border: 1px solid #ccc;
  }
  .table-wrapper {
    flex: 8;
  }
  .filter-wrapper {
    padding: 1%;
    flex: 2;
  }
}
