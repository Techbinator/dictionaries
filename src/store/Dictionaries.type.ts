export interface IDictionary {
  uuid: string;
  name: string;
  translationsUUID: string[];
}

export interface ITranslation {
  uuid: string;
  domain: string;
  range: string;
}

export interface IDictionaries {
  [uuid: string]: IDictionary;
}

export interface ITranslations {
  [uuid: string]: ITranslation;
}
export type TeditTranslation = (translation: ITranslation) => void;
export type TaddTranslation = (
  dictionaryUUID: string,
  translation: ITranslation
) => void;
export type TremoveTranslation = (
  dictionaryUUID: string,
  translationUUID: string
) => void;
export type TaddDictionary = (name: string) => void;
export type TremoveDictionary = (uuid: string) => void;
export type TactivateDictionary = (uuid: string) => void;
export interface IDictionaryProvider {
  dictionaries: IDictionaries;
  translations: ITranslations;
  addDictionary: TaddDictionary;
  activeDictionary: string;
  removeDictionary: TremoveDictionary;
  removeTranslation: TremoveTranslation;
  addTranslation: TaddTranslation;
  editTranslation: TeditTranslation;
  activateDictionary: TactivateDictionary;
}
