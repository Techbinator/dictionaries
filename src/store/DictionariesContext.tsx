import React from "react";
import {
  IDictionaryProvider,
  IDictionaries,
  ITranslation,
  ITranslations
} from "./Dictionaries.type";
import { v1 } from "uuid";

export const DictionariesContext = React.createContext({});
export class DictionariesProvider extends React.Component {
  addDictionary = (name: string) => {
    const uuid = v1();
    const newDictionary = {
      uuid,
      name,
      translationsUUID: []
    };
    this.setState({
      dictionaries: { [uuid]: newDictionary, ...this.state.dictionaries }
    });
  };

  activateDictionary = (dictionaryUUID: string) => {
    this.setState({ activeDictionary: dictionaryUUID });
  };

  removeDictionary = (uuid: string) => {
    const { activeDictionary, dictionaries, translations } = this.state;
    const dictionary = dictionaries[uuid];
    const newDictionaries: IDictionaries = {
      ...dictionaries
    };
    const NewTranslations = {
      ...translations
    };
    dictionary.translationsUUID.forEach(uuid => delete NewTranslations[uuid]);

    delete newDictionaries[uuid];
    this.setState({
      dictionaries: newDictionaries,
      activeDictionary: activeDictionary === uuid ? "" : activeDictionary,
      translations: NewTranslations
    });
  };

  removeTranslation = (dictionaryUUID: string, translationUUID: string) => {
    const { dictionaries, translations } = this.state;
    const newTranslations: ITranslations = {
      ...translations
    };
    delete newTranslations[translationUUID];
    this.setState({
      translations: newTranslations,
      dictionaries: {
        ...dictionaries,
        [dictionaryUUID]: {
          ...dictionaries[dictionaryUUID],
          translationsUUID: dictionaries[
            dictionaryUUID
          ].translationsUUID.filter(tuuid => tuuid !== translationUUID)
        }
      }
    });
  };

  addTranslation = (dictionaryUUID: string, translation: ITranslation) => {
    const { dictionaries, translations } = this.state;
    const uuid = v1();
    const newTranslation = {
      ...translation,
      uuid
    };

    const dictionaryTranslations = dictionaries[dictionaryUUID];
    this.setState({
      translations: {
        [uuid]: newTranslation,
        ...translations
      },
      dictionaries: {
        ...dictionaries,
        [dictionaryUUID]: {
          ...dictionaryTranslations,
          translationsUUID: [uuid, ...dictionaryTranslations.translationsUUID]
        }
      }
    });
  };

  editTranslation = (translation: ITranslation) => {
    this.setState({
      translations: {
        ...this.state.translations,
        [translation.uuid]: translation
      }
    });
  };

  state: IDictionaryProvider = {
    dictionaries: {},
    translations: {},
    activeDictionary: "",
    addDictionary: this.addDictionary,
    removeDictionary: this.removeDictionary,
    removeTranslation: this.removeTranslation,
    addTranslation: this.addTranslation,
    editTranslation: this.editTranslation,
    activateDictionary: this.activateDictionary
  };

  //get from localstorage
  componentWillMount() {
    const dictionaries = localStorage.getItem("dictionaries");
    if (dictionaries) {
      this.setState({ dictionaries: JSON.parse(dictionaries) });
    }

    const translations = localStorage.getItem("translations");
    if (translations) {
      this.setState({ translations: JSON.parse(translations) });
    }
  }
  //save to localstorage
  componentDidUpdate({}, prevState: IDictionaryProvider) {
    const { dictionaries, translations } = this.state;
    if (
      JSON.stringify(dictionaries) !== JSON.stringify(prevState.dictionaries)
    ) {
      // save to localstorage
      localStorage.setItem("dictionaries", JSON.stringify(dictionaries));
    }
    if (
      JSON.stringify(translations) !== JSON.stringify(prevState.translations)
    ) {
      // save to localstorage
      localStorage.setItem("translations", JSON.stringify(translations));
    }
  }

  render() {
    return (
      <DictionariesContext.Provider value={this.state}>
        {this.props.children}
      </DictionariesContext.Provider>
    );
  }
}
