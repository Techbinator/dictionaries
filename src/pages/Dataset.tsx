import React from "react";
import RadioGroupToggle from "../components/RadioGroupToggle";
import DatasetHead from "../components/DatasetHead";
import DatasetBody from "../components/DatasetBody";
import { DictionariesContext } from "../store/DictionariesContext";

import "./Dataset.scss";
import dataset from "../helpers/__mocks__/dataset.json";

export default class Dataset extends React.Component {
  render() {
    const {
      dictionaries,
      activeDictionary,
      activateDictionary,
      translations
    } = this.context;

    return (
      <div className="dataset one-edge-shadow">
        <div className="table-wrapper">
          <table>
            <DatasetHead {...dataset[0]} />
            <DatasetBody
              data={dataset}
              dictionary={dictionaries[activeDictionary]}
              translations={translations}
            />
          </table>
        </div>
        <div className="filter-wrapper">
          <RadioGroupToggle
            dictionaries={dictionaries}
            activeDictionary={activeDictionary}
            activateDictionary={activateDictionary}
          />
        </div>
      </div>
    );
  }
}

Dataset.contextType = DictionariesContext;
