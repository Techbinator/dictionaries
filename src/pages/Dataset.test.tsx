import React from "react";
import Dataset from "./Dataset";
import { mount } from "enzyme";
import { DictionariesProvider } from "../store/DictionariesContext";
import contextMock from "../helpers/__mocks__/context.mock";
jest.mock(
  "../helpers/__mocks__/dataset.json",
  () => require("../helpers/__mocks__/smallDataset.json"),
  {
    virtual: true
  }
);
describe("Dataset component", () => {
  it("should render snapshot with default values", () => {
    const wrapper = mount(
      <DictionariesProvider {...contextMock}>
        <Dataset />
      </DictionariesProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
