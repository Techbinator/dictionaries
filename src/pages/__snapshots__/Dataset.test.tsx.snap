// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dataset component should render snapshot with default values 1`] = `
<DictionariesProvider
  activateDictionary={[MockFunction]}
  activeDictionary="619ce930-4116-11e9-9306-ef359c18301d"
  addDictionary={[MockFunction]}
  addTranslation={[MockFunction]}
  dictionaries={
    Object {
      "4aa2cce0-4116-11e9-9306-ef359c18301d": Object {
        "name": "test",
        "translationsUUID": Array [
          "58f6b540-4116-11e9-9306-ef359c18301d",
          "55a62ab0-4116-11e9-9306-ef359c18301d",
        ],
        "uuid": "4aa2cce0-4116-11e9-9306-ef359c18301d",
      },
      "619ce930-4116-11e9-9306-ef359c18301d": Object {
        "name": "dictionary2",
        "translationsUUID": Array [
          "6e270e10-4116-11e9-9306-ef359c18301d",
        ],
        "uuid": "619ce930-4116-11e9-9306-ef359c18301d",
      },
    }
  }
  editTranslation={[MockFunction]}
  removeDictionary={[MockFunction]}
  removeTranslation={[MockFunction]}
  translations={
    Object {
      "55a62ab0-4116-11e9-9306-ef359c18301d": Object {
        "domain": "Crimson",
        "range": "newText",
        "uuid": "55a62ab0-4116-11e9-9306-ef359c18301d",
      },
      "58f6b540-4116-11e9-9306-ef359c18301d": Object {
        "domain": "a",
        "range": "b",
        "uuid": "58f6b540-4116-11e9-9306-ef359c18301d",
      },
      "6e270e10-4116-11e9-9306-ef359c18301d": Object {
        "domain": "Turquoise",
        "range": "newRange",
        "uuid": "6e270e10-4116-11e9-9306-ef359c18301d",
      },
    }
  }
>
  <Dataset>
    <div
      className="dataset one-edge-shadow"
    >
      <div
        className="table-wrapper"
      >
        <table>
          <_default
            car="GMC"
            city="Zhvanets"
            color="Turquoise"
            id={1}
            payment="jcb"
            price={30}
            product="Sage Ground Wiberg"
          >
            <thead>
              <tr>
                <th
                  key="id"
                >
                  id
                </th>
                <th
                  key="product"
                >
                  product
                </th>
                <th
                  key="color"
                >
                  color
                </th>
                <th
                  key="price"
                >
                  price
                </th>
                <th
                  key="car"
                >
                  car
                </th>
                <th
                  key="payment"
                >
                  payment
                </th>
                <th
                  key="city"
                >
                  city
                </th>
              </tr>
            </thead>
          </_default>
          <_default
            data={
              Array [
                Object {
                  "car": "GMC",
                  "city": "Zhvanets",
                  "color": "Turquoise",
                  "id": 1,
                  "payment": "jcb",
                  "price": 30,
                  "product": "Sage Ground Wiberg",
                },
                Object {
                  "car": "Suzuki",
                  "city": "Mberubu",
                  "color": "Crimson",
                  "id": 2,
                  "payment": "diners-club-international",
                  "price": 69,
                  "product": "Pizza Pizza Dough",
                },
                Object {
                  "car": "Mercury",
                  "city": "Kedungsumurkrajan",
                  "color": "Blue",
                  "id": 3,
                  "payment": "diners-club-us-ca",
                  "price": 50,
                  "product": "Cheese - Gorgonzola",
                },
                Object {
                  "car": "Volvo",
                  "city": "DoÃ±a Remedios Trinidad",
                  "color": "Orange",
                  "id": 4,
                  "payment": "jcb",
                  "price": 65,
                  "product": "Coconut - Whole",
                },
              ]
            }
            translations={Object {}}
          >
            <tbody>
              <tr
                key="0"
              >
                <_default
                  domains={Array []}
                  key="1"
                  text={1}
                  translations={Object {}}
                >
                  <td>
                    1
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Sage Ground Wiberg"
                  text="Sage Ground Wiberg"
                  translations={Object {}}
                >
                  <td>
                    Sage Ground Wiberg
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Turquoise"
                  text="Turquoise"
                  translations={Object {}}
                >
                  <td>
                    Turquoise
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="30"
                  text={30}
                  translations={Object {}}
                >
                  <td>
                    30
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="GMC"
                  text="GMC"
                  translations={Object {}}
                >
                  <td>
                    GMC
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="jcb"
                  text="jcb"
                  translations={Object {}}
                >
                  <td>
                    jcb
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Zhvanets"
                  text="Zhvanets"
                  translations={Object {}}
                >
                  <td>
                    Zhvanets
                  </td>
                </_default>
              </tr>
              <tr
                key="1"
              >
                <_default
                  domains={Array []}
                  key="2"
                  text={2}
                  translations={Object {}}
                >
                  <td>
                    2
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Pizza Pizza Dough"
                  text="Pizza Pizza Dough"
                  translations={Object {}}
                >
                  <td>
                    Pizza Pizza Dough
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Crimson"
                  text="Crimson"
                  translations={Object {}}
                >
                  <td>
                    Crimson
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="69"
                  text={69}
                  translations={Object {}}
                >
                  <td>
                    69
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Suzuki"
                  text="Suzuki"
                  translations={Object {}}
                >
                  <td>
                    Suzuki
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="diners-club-international"
                  text="diners-club-international"
                  translations={Object {}}
                >
                  <td>
                    diners-club-international
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Mberubu"
                  text="Mberubu"
                  translations={Object {}}
                >
                  <td>
                    Mberubu
                  </td>
                </_default>
              </tr>
              <tr
                key="2"
              >
                <_default
                  domains={Array []}
                  key="3"
                  text={3}
                  translations={Object {}}
                >
                  <td>
                    3
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Cheese - Gorgonzola"
                  text="Cheese - Gorgonzola"
                  translations={Object {}}
                >
                  <td>
                    Cheese - Gorgonzola
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Blue"
                  text="Blue"
                  translations={Object {}}
                >
                  <td>
                    Blue
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="50"
                  text={50}
                  translations={Object {}}
                >
                  <td>
                    50
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Mercury"
                  text="Mercury"
                  translations={Object {}}
                >
                  <td>
                    Mercury
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="diners-club-us-ca"
                  text="diners-club-us-ca"
                  translations={Object {}}
                >
                  <td>
                    diners-club-us-ca
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Kedungsumurkrajan"
                  text="Kedungsumurkrajan"
                  translations={Object {}}
                >
                  <td>
                    Kedungsumurkrajan
                  </td>
                </_default>
              </tr>
              <tr
                key="3"
              >
                <_default
                  domains={Array []}
                  key="4"
                  text={4}
                  translations={Object {}}
                >
                  <td>
                    4
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Coconut - Whole"
                  text="Coconut - Whole"
                  translations={Object {}}
                >
                  <td>
                    Coconut - Whole
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Orange"
                  text="Orange"
                  translations={Object {}}
                >
                  <td>
                    Orange
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="65"
                  text={65}
                  translations={Object {}}
                >
                  <td>
                    65
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="Volvo"
                  text="Volvo"
                  translations={Object {}}
                >
                  <td>
                    Volvo
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="jcb"
                  text="jcb"
                  translations={Object {}}
                >
                  <td>
                    jcb
                  </td>
                </_default>
                <_default
                  domains={Array []}
                  key="DoÃ±a Remedios Trinidad"
                  text="DoÃ±a Remedios Trinidad"
                  translations={Object {}}
                >
                  <td>
                    DoÃ±a Remedios Trinidad
                  </td>
                </_default>
              </tr>
            </tbody>
          </_default>
        </table>
      </div>
      <div
        className="filter-wrapper"
      >
        <RadioToggle
          activateDictionary={[Function]}
          activeDictionary=""
          dictionaries={Object {}}
        >
          <div
            className="radio-group-container"
          >
            <input
              checked={true}
              id="radio-0"
              name="dictionary-selection"
              readOnly={true}
              type="radio"
            />
            <label
              htmlFor="radio-0"
              onClick={[Function]}
            >
              Initial data
            </label>
          </div>
        </RadioToggle>
      </div>
    </div>
  </Dataset>
</DictionariesProvider>
`;
